### PR TITLE
inductor: tigher upperbound for rblock scaling

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109839

Previously when we deciding if dynamically scaling down rblock, we use the following formule to compute the upper bound of number of blocks per sm:
```
max_threads_per_multi_processo / (32 * num_warps)
```

This is correct but it's a bit loose and some times because of the loose upper bound, we skip some optimization opportunities.

The new upper bound is: 65536 / n_reg_used_by_each_block . This is a tighter upper bound and can be helpful if the kernel uses too many registers (i.e. much larger than 32).

For kernel https://gist.github.com/shunting314/59aeafd297ed8ff03aa12030a2dd41ae (this is a real kernel inductor generates for HF), the change improve its perf from:
0.485ms    0.332GB    684.29GB/s
to 
0.240ms    0.332GB    1382.70GB/s

. The perf is bad previsouly because of register spills

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_inductor/triton_heuristics.py [hunk 1]
```diff
@@ -198,6 +199,10 @@ def precompile(self, warm_cache_only_with_cc=None):
                     if nreg is None:
                         continue
 
+                    # make sure rblock is not too small
+                    if rblock <= 64:
+                        continue
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/triton_heuristics.py:204]: Generally good to have simple checks first so later code can assume rblock > 64

File: torch/_inductor/triton_heuristics.py [hunk 2]
```diff
@@ -211,17 +216,26 @@ def precompile(self, warm_cache_only_with_cc=None):
...
+                    # Previously we set max_blocks_per_sm to 'max_threads_per_multi_processo / (32 * num_warps)'
+                    # The formula below is a tighter upper bound since we have the assumption that
+                    #   nreg > 65536 // device_prop.max_threads_per_multi_processor
+                    # due to the if condition above and:
+                    #   65536 / nreg_per_block
+                    #   = 65536 / (nreg * 32 * num_warps)
+                    #   < 65536 / ((65536 / max_threads_per_multi_processor) * 32 * num_warps)
+                    #   = max_threads_per_multi_processor / (32 * num_warps)
+                    # Using a tigher upper bound can reveal more optimization opportunities.
+                    max_blocks_per_sm = max(65536 // nreg_per_block, 1)
```
[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/triton_heuristics.py:232]: > Here, if 65536 // nreg_per_block < 1 there is going to be register spillage, so we probably want to decrease the RBLOCK as well.

So I use `max(65536 // nreg_per_block, 1)` just to be safe. nreg_per_block actually is the register usage after considering register spills. The spilled registers are not counted in nreg_per_block . So if `65536 // nreg_per_block < 1` probably the kernel will not able to be run. 

Register spill is implicitly handled here actually. When register spills, each thread gonna uses 255 registers (the maximum # of register a thread can use). In that case, we have large chance to trigger the decreasing or RBLOCK.

[the remark this answers is not in the document]
Comment (shunting314) [reply torch/_inductor/triton_heuristics.py:232]: > max_blocks_per_sm = min(65536 // nreg_per_block, max_warps_per_sm // triton_conf.num_warps)

This is equivalent to the current code. My comment in the code shows why the following is true:
```
65536 // nreg_per_block < max_warps_per_sm // triton_conf.num_warps
```
here.

Scaling up RBLOCK can do harm since we increase register usage. Even if there is no register spills,  large register usage can still hurt occupancy and may result in under-performing configs.

**Comment by shunting314** [comment]
Here is the perf test result: [link](https://hud.pytorch.org/benchmark/compilers?startTime=Fri%2C%2015%20Sep%202023%2017%3A53%3A18%20GMT&stopTime=Fri%2C%2022%20Sep%202023%2017%3A53%3A18%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=shunting-tigher-upperbound&lCommit=ee9e5a542bc7cd589f4bb66c12bd6e8dbb8c9074&rBranch=main&rCommit=2c1554a0323107d821be3ff13df7833b9f0b960d)
- the compilation time is overall neutral
- I see 1% perf increase in TIMM but since I see no green cells, I'll consider it as noise.
- I see 1% perf decrease in HF. But the perf should not cause real perf degradation since the net effect is we consider 1 more triton config in some cases. Actually I see no red cells but 1 green cells for HF: DebertaV2ForQuestionAnswering increase from 2.44 to 2.62. Also the model that the spilled kernel comes from is BlenderbotForCausalLM . BlenderbotForCausalLM also increase from 1.17x to 1.18x.

So the PR improve things in general but overall metric is neutral.

File: torch/_inductor/triton_heuristics.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lezcano) [reply torch/_inductor/triton_heuristics.py:232]: > When register spills, each thread gonna uses 255 registers (the maximum # of register a thread can use). In that case, we have large chance to trigger the decreasing of RBLOCK.

Why don't we handle it explicitly now that we are at it?
